doc/rtd/development/datasource_creation.rst

```diff
@@ -0,0 +1,89 @@
+.. _datasource_creation:
+
+Datasource creation
+*******************
+
+Since there are multiple ways to provide `user data`, `meta data`, and
+`vendor data` (each cloud solution seems to prefer its own way), a datasource
+abstract class was internally created to allow for a single way to access the
+different cloud systems methods, providing this data through the typical usage
+of subclasses. See :file:`cloud-init/sources/__init__.py` as a starting
+point for understanding the datastructures used.
+
+The datasource objects have a few touch points with ``cloud-init``. If you
```

Comment:
Is a datasource object the same thing as the datasource class mentioned above?

Reply:
Good question.

There is a relationship between the [(abstract) parent class](https://github.com/canonical/cloud-init/blob/34637a49df13a054c4355de675b7cf286450b2f5/cloudinit/sources/__init__.py#L165) and the datasource object. The various child classes are defined in `cloudinit/sources/DataSource*.py` and inherit from `cloudinit.sources.DataSource`

As for this statement:

> The datasource objects have a few touch points with ``cloud-init``.

It does feel odd that we're trying to say that "there are several things that need to be updated when adding a new datasource", yet we use the word `object`, which only refers to only an instantance of the child class, not the creation of the child class itself. This wording seems odd to me, since adding a new datasource will always involve defining a new class, and is unlikely to require manipulating how the datasource object gets used.  Perhaps this would be more clear if it just said:

> Multiple parts of the codebase require an update when adding a new datasource.

Thoughts?

Reply:
The reason why I asked is because it reads like we're introducing a new term or concept (in "database objects") that hasn't been defined in the page. I know objects usually have a specific meaning, so I don't think it hurts us to explicitly state that relationship you just described, unless (since it's a devdoc page) you think our readers will already understand all of that?

I do agree that stating "Multiple parts of the codebase require an update when adding a new datasource." is clearer than what was there originally, but should we also state explicitly which parts of the codebase? (or is that fully taken care of in the list that follows?)